manta-pay/src/crypto/poseidon/encryption.rs

```diff
@@ -0,0 +1,191 @@
+// Copyright 2019-2022 Manta Network.
+// This file is part of manta-rs.
+//
+// manta-rs is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// manta-rs is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with manta-rs.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Poseidon Permutation Implementation
+
+use crate::crypto::poseidon::{Field, Permutation, Specification, State};
+use alloc::{boxed::Box, vec::Vec};
+use core::{fmt::Debug, hash::Hash, iter, marker::PhantomData, mem, slice};
+use manta_crypto::{
+    permutation::{
+        duplex::{Setup, Types, Verify},
+        sponge::{Read, Write},
+        PseudorandomPermutation,
+    },
+    rand::{Rand, RngCore, Sample},
+};
+
+#[cfg(feature = "serde")]
+use manta_util::serde::{Deserialize, Serialize};
+
+///
+pub struct SetupBlock<S, COM = ()>(PhantomData<(S, COM)>)
+where
+    S: Specification<COM>;
+
+impl<S, COM> Write<Permutation<S, COM>, COM> for SetupBlock<S, COM>
+where
+    S: Specification<COM>,
+{
+    type Output = ();
+
+    #[inline]
+    fn write(&self, state: &mut State<S, COM>, compiler: &mut COM) -> Self::Output {
+        todo!()
+    }
+}
+
+///
+pub struct PlaintextBlock<S, COM = ()>(Box<[S::Field]>)
+where
+    S: Specification<COM>;
+
+impl<S, COM> Write<Permutation<S, COM>, COM> for PlaintextBlock<S, COM>
+where
+    S: Specification<COM>,
+    S::Field: Clone,
+{
+    type Output = CiphertextBlock<S, COM>;
+
+    #[inline]
+    fn write(&self, state: &mut State<S, COM>, compiler: &mut COM) -> Self::Output {
+        for (i, elem) in state.iter_mut().skip(1).enumerate() {
+            *elem = S::add(elem, &self.0[i], compiler);
+        }
+        CiphertextBlock(state.iter().skip(1).cloned().collect())
```

Comment:
I'm not sure if this is the correct ciphertext block. In the paper, we XOR/ADD the plaintext to the state (correspond to the for-loop here), do a permute, and the ciphertext is plaintext XOR/ADD the state after permute. However, the permutation is missing here? Correct me if wrong

Reply:
(Reading line numbers from the paper).

On line 15 we start by first computing the ciphertext block against the current state (`Z` is the current state without capacity) then we call `duplexing` against `B_0` which again XORs with the state (without capacity). (We are not using frame bits in our algorithm, but the `duplex` API supports frame bits if people want them). So instead we collapse these two operations into the API called `write` which produces the ciphertext block `C_0` and updates the state (the first half of `duplexing`). Then, we immediately permute the state in the `duplex_encryption` function which is the second-half of the `duplexing` call on line 17. Then we loop at that point instead of looping on line 18. We shift the loop by one instruction which still gives the same result (line 15 becomes part of the first loop instead of some instruction before the loop starts and line 20 is also just part of the loop, the last instruction in the last iteration).